# Task 1: Data Quality Assessment

Read the xlsx to pandas

Extract Sheet Names

In [1]:
import pandas as pd
import numpy as np

sheets = pd.ExcelFile('KPMG_VI_New_raw_data_update_final.xlsx')
sheets.sheet_names

['Title Sheet',
 'Transactions',
 'NewCustomerList',
 'CustomerDemographic',
 'CustomerAddress']

Read into pandas dataframe

In [2]:
# First rows of the sheets are comments, not part of dataframes
df_transactions = pd.read_excel('KPMG_VI_New_raw_data_update_final.xlsx', sheet_name='Transactions', skiprows=1)
df_new_customer_list = pd.read_excel('KPMG_VI_New_raw_data_update_final.xlsx', sheet_name='NewCustomerList', skiprows=1)
df_customer_demographic = pd.read_excel('KPMG_VI_New_raw_data_update_final.xlsx', sheet_name='CustomerDemographic', skiprows=1)
df_customer_address = pd.read_excel('KPMG_VI_New_raw_data_update_final.xlsx', sheet_name='CustomerAddress', skiprows=1)

## Assess Transactions

In [3]:
df_transactions.head()

transaction_id  product_id  customer_id transaction_date  online_order  \
0               1           2         2950       2017-02-25           0.0   
1               2           3         3120       2017-05-21           1.0   
2               3          37          402       2017-10-16           0.0   
3               4          88         3135       2017-08-31           0.0   
4               5          78          787       2017-10-01           1.0   

  order_status           brand product_line product_class product_size  \
0     Approved           Solex     Standard        medium       medium   
1     Approved   Trek Bicycles     Standard        medium        large   
2     Approved      OHM Cycles     Standard           low       medium   
3     Approved  Norco Bicycles     Standard        medium       medium   
4     Approved  Giant Bicycles     Standard        medium        large   

   list_price  standard_cost  product_first_sold_date  
0       71.49          53.62                  41245.0  
1     2091.47         388.92                  41701.0  
2     1793.43         248.82                  36361.0  
3     1198.46         381.10                  36145.0  
4     1765.30         709.48                  42226.0

### Correct Values - Accuracy

In [4]:
df_transactions.describe()

transaction_id   product_id   customer_id  online_order    list_price  \
count    20000.000000  20000.00000  20000.000000  19640.000000  20000.000000   
mean     10000.500000     45.36465   1738.246050      0.500458   1107.829449   
std       5773.647028     30.75359   1011.951046      0.500013    582.825242   
min          1.000000      0.00000      1.000000      0.000000     12.010000   
25%       5000.750000     18.00000    857.750000      0.000000    575.270000   
50%      10000.500000     44.00000   1736.000000      1.000000   1163.890000   
75%      15000.250000     72.00000   2613.000000      1.000000   1635.300000   
max      20000.000000    100.00000   5034.000000      1.000000   2091.470000   

       standard_cost  product_first_sold_date  
count   19803.000000             19803.000000  
mean      556.046951             38199.776549  
std       405.955660              2875.201110  
min         7.210000             33259.000000  
25%       215.140000             35667.000000  
50%       507.580000             38216.000000  
75%       795.100000             40672.000000  
max      1759.850000             42710.000000

We see that the mininum list_price is 12.01, which does not really make sense to a normal bicycle. The minimum standard_cost is 7.21, which does not really make sense to a normal bicycle either. Now let's explore these cases.

In [5]:
np.sort(df_transactions['list_price'].unique())

array([  12.01,   16.08,   26.15,   32.44,   36.78,   56.21,   60.34,
         60.72,   71.16,   71.49,   74.89,  100.35,  101.55,  122.74,
        126.36,  149.3 ,  150.51,  156.56,  162.87,  172.09,  175.89,
        183.86,  202.62,  205.84,  226.94,  227.88,  230.91,  235.63,
        237.44,  240.72,  290.62,  294.35,  310.57,  311.54,  311.57,
        317.3 ,  326.86,  330.36,  330.91,  356.05,  356.39,  358.39,
        360.4 ,  363.01,  366.22,  398.08,  402.14,  405.94,  416.98,
        435.66,  437.51,  441.49,  447.25,  465.71,  470.07,  478.16,
        483.12,  487.8 ,  489.58,  494.96,  495.72,  499.53,  517.53,
        533.51,  541.44,  543.39,  544.05,  554.38,  569.56,  571.27,
        574.64,  575.27,  586.45,  587.02,  590.26,  605.54,  634.57,
        638.29,  639.38,  642.31,  642.7 ,  643.95,  675.28,  678.25,
        687.9 ,  688.63,  695.23,  710.55,  710.59,  724.37,  731.41,
        742.54,  744.54,  748.17,  750.25,  752.36,  752.64,  753.01,
        753.76,  756

In [6]:
df_transactions[df_transactions['list_price'] == 12.01].sample(5)

transaction_id  product_id  customer_id transaction_date  online_order  \
7071            7072           0          661       2017-07-16           1.0   
5555            5556           0         1323       2017-06-06           0.0   
9583            9584          19         1545       2017-06-17           1.0   
3142            3143          19          479       2017-08-10           0.0   
5078            5079          19          871       2017-10-27           1.0   

     order_status       brand product_line product_class product_size  \
7071     Approved  OHM Cycles         Road          high        large   
5555     Approved  OHM Cycles         Road          high        large   
9583     Approved  OHM Cycles         Road          high        large   
3142     Approved  OHM Cycles         Road          high        large   
5078     Approved  OHM Cycles         Road          high        large   

      list_price  standard_cost  product_first_sold_date  
7071       12.01           7.21                  39880.0  
5555       12.01           7.21                  35160.0  
9583       12.01           7.21                  34165.0  
3142       12.01           7.21                  36334.0  
5078       12.01           7.21                  40670.0

We can see that although the list_price is very low, the standard cost is even lower thatn the list_price. This makes sense.

### Data Fields with Values - Completeness

In [7]:
df_transactions.isnull().sum()

transaction_id               0
product_id                   0
customer_id                  0
transaction_date             0
online_order               360
order_status                 0
brand                      197
product_line               197
product_class              197
product_size               197
list_price                   0
standard_cost              197
product_first_sold_date    197
dtype: int64

We have multiple rows with null values. This need to be addressed with the client.

### Values Free from Contradition - Consistency

In [8]:
df_transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           20000 non-null  int64         
 1   product_id               20000 non-null  int64         
 2   customer_id              20000 non-null  int64         
 3   transaction_date         20000 non-null  datetime64[ns]
 4   online_order             19640 non-null  float64       
 5   order_status             20000 non-null  object        
 6   brand                    19803 non-null  object        
 7   product_line             19803 non-null  object        
 8   product_class            19803 non-null  object        
 9   product_size             19803 non-null  object        
 10  list_price               20000 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_date  19803 n

order_status, brand, product_line, product_class, and product_size are object datatypes. Let's explore them.

In [9]:
df_transactions['order_status'].unique()

array(['Approved', 'Cancelled'], dtype=object)

In [10]:
df_transactions['brand'].unique()

array(['Solex', 'Trek Bicycles', 'OHM Cycles', 'Norco Bicycles',
       'Giant Bicycles', 'WeareA2B', nan], dtype=object)

In [11]:
df_transactions['product_line'].unique()

array(['Standard', 'Road', 'Mountain', 'Touring', nan], dtype=object)

In [12]:
df_transactions['product_class'].unique()

array(['medium', 'low', 'high', nan], dtype=object)

In [13]:
df_transactions['product_size'].unique()

array(['medium', 'large', 'small', nan], dtype=object)

We can see the data is consistent.

### Data Containing Allowable Values - Validity

In [14]:
df_transactions.describe()

transaction_id   product_id   customer_id  online_order    list_price  \
count    20000.000000  20000.00000  20000.000000  19640.000000  20000.000000   
mean     10000.500000     45.36465   1738.246050      0.500458   1107.829449   
std       5773.647028     30.75359   1011.951046      0.500013    582.825242   
min          1.000000      0.00000      1.000000      0.000000     12.010000   
25%       5000.750000     18.00000    857.750000      0.000000    575.270000   
50%      10000.500000     44.00000   1736.000000      1.000000   1163.890000   
75%      15000.250000     72.00000   2613.000000      1.000000   1635.300000   
max      20000.000000    100.00000   5034.000000      1.000000   2091.470000   

       standard_cost  product_first_sold_date  
count   19803.000000             19803.000000  
mean      556.046951             38199.776549  
std       405.955660              2875.201110  
min         7.210000             33259.000000  
25%       215.140000             35667.000000  
50%       507.580000             38216.000000  
75%       795.100000             40672.000000  
max      1759.850000             42710.000000

In [15]:
df_transactions['transaction_date'].sort_values()

516     2017-01-01
5876    2017-01-01
3459    2017-01-01
12484   2017-01-01
19130   2017-01-01
           ...    
605     2017-12-30
15269   2017-12-30
19906   2017-12-30
15756   2017-12-30
12003   2017-12-30
Name: transaction_date, Length: 20000, dtype: datetime64[ns]

Data ensures good validity.

### Records that are Duplicated - Uniqueness

In [16]:
df_transactions[df_transactions['transaction_id'].duplicated()]

Empty DataFrame
Columns: [transaction_id, product_id, customer_id, transaction_date, online_order, order_status, brand, product_line, product_class, product_size, list_price, standard_cost, product_first_sold_date]
Index: []

Data ensures good uniqueness

## Assess NewCustomerList

### Correct Values - Accuracy

In [17]:
df_new_customer_list.head()

first_name  last_name  gender  past_3_years_bike_related_purchases  \
0    Chickie    Brister    Male                                   86   
1      Morly     Genery    Male                                   69   
2    Ardelis  Forrester  Female                                   10   
3     Lucine      Stutt  Female                                   64   
4    Melinda     Hadlee  Female                                   34   

         DOB                   job_title job_industry_category  \
0 1957-07-12             General Manager         Manufacturing   
1 1970-03-22         Structural Engineer              Property   
2 1974-08-28      Senior Cost Accountant    Financial Services   
3 1979-01-28  Account Representative III         Manufacturing   
4 1965-09-21           Financial Analyst    Financial Services   

      wealth_segment deceased_indicator owns_car  ...  state    country  \
0      Mass Customer                  N      Yes  ...    QLD  Australia   
1      Mass Customer                  N       No  ...    NSW  Australia   
2  Affluent Customer                  N       No  ...    VIC  Australia   
3  Affluent Customer                  N      Yes  ...    QLD  Australia   
4  Affluent Customer                  N       No  ...    NSW  Australia   

   property_valuation Unnamed: 16 Unnamed: 17  Unnamed: 18  Unnamed: 19  \
0                   6        0.56      0.7000       0.8750     0.743750   
1                  11        0.89      0.8900       1.1125     0.945625   
2                   5        1.01      1.0100       1.0100     1.010000   
3                   1        0.87      1.0875       1.0875     1.087500   
4                   9        0.52      0.5200       0.6500     0.650000   

   Unnamed: 20  Rank     Value  
0            1     1  1.718750  
1            1     1  1.718750  
2            1     1  1.718750  
3            4     4  1.703125  
4            4     4  1.703125  

[5 rows x 23 columns]

In [18]:
df_new_customer_list.describe()

past_3_years_bike_related_purchases       tenure     postcode  \
count                          1000.000000  1000.000000  1000.000000   
mean                             49.836000    11.388000  3019.227000   
std                              27.796686     5.037145   848.895767   
min                               0.000000     0.000000  2000.000000   
25%                              26.750000     7.000000  2209.000000   
50%                              51.000000    11.000000  2800.000000   
75%                              72.000000    15.000000  3845.500000   
max                              99.000000    22.000000  4879.000000   

       property_valuation  Unnamed: 16  Unnamed: 17  Unnamed: 18  Unnamed: 19  \
count         1000.000000  1000.000000  1000.000000  1000.000000  1000.000000   
mean             7.397000     0.747340     0.839005     0.942673     0.870514   
std              2.758804     0.205082     0.248858     0.294832     0.280891   
min              1.000000     0.400000     0.400000     0.400000     0.340000   
25%              6.000000     0.570000     0.637500     0.712500     0.658750   
50%              8.000000     0.750000     0.820000     0.912500     0.842625   
75%              9.000000     0.920000     1.031875     1.142969     1.062500   
max             12.000000     1.100000     1.375000     1.718750     1.718750   

       Unnamed: 20         Rank        Value  
count  1000.000000  1000.000000  1000.000000  
mean    498.819000   498.819000     0.881714  
std     288.810997   288.810997     0.293525  
min       1.000000     1.000000     0.340000  
25%     250.000000   250.000000     0.649531  
50%     500.000000   500.000000     0.860000  
75%     750.250000   750.250000     1.075000  
max    1000.000000  1000.000000     1.718750

In [19]:
df_new_customer_list.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 23 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   first_name                           1000 non-null   object        
 1   last_name                            971 non-null    object        
 2   gender                               1000 non-null   object        
 3   past_3_years_bike_related_purchases  1000 non-null   int64         
 4   DOB                                  983 non-null    datetime64[ns]
 5   job_title                            894 non-null    object        
 6   job_industry_category                835 non-null    object        
 7   wealth_segment                       1000 non-null   object        
 8   deceased_indicator                   1000 non-null   object        
 9   owns_car                             1000 non-null   object        
 10  tenure       

In [20]:
df_new_customer_list['DOB'].sort_values()

797   1938-06-08
885   1938-06-09
769   1938-08-05
585   1938-08-30
265   1938-08-31
         ...    
775          NaT
835          NaT
883          NaT
904          NaT
984          NaT
Name: DOB, Length: 1000, dtype: datetime64[ns]

In [21]:
df_new_customer_list['DOB'].sort_values(ascending=False)

354   2002-02-27
396   2002-01-17
763   2002-01-04
475   2001-12-29
996   2001-10-17
         ...    
775          NaT
835          NaT
883          NaT
904          NaT
984          NaT
Name: DOB, Length: 1000, dtype: datetime64[ns]

Min, max values, and date of birth make sense. Data ensures good accuracy.

### Data Fields with Values - Completeness

In [22]:
df_new_customer_list.isnull().sum()

first_name                               0
last_name                               29
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     17
job_title                              106
job_industry_category                  165
wealth_segment                           0
deceased_indicator                       0
owns_car                                 0
tenure                                   0
address                                  0
postcode                                 0
state                                    0
country                                  0
property_valuation                       0
Unnamed: 16                              0
Unnamed: 17                              0
Unnamed: 18                              0
Unnamed: 19                              0
Unnamed: 20                              0
Rank                                     0
Value                                    0
dtype: int6

We have multiple rows with null values. This need to be addressed with the client.

### Values Free from Contradition - Consistency

In [23]:
df_new_customer_list.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 23 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   first_name                           1000 non-null   object        
 1   last_name                            971 non-null    object        
 2   gender                               1000 non-null   object        
 3   past_3_years_bike_related_purchases  1000 non-null   int64         
 4   DOB                                  983 non-null    datetime64[ns]
 5   job_title                            894 non-null    object        
 6   job_industry_category                835 non-null    object        
 7   wealth_segment                       1000 non-null   object        
 8   deceased_indicator                   1000 non-null   object        
 9   owns_car                             1000 non-null   object        
 10  tenure       

Columns with dtype of "object" need to be assessed.

In [24]:
df_new_customer_list['gender'].unique()

array(['Male', 'Female', 'U'], dtype=object)

In [25]:
df_new_customer_list['job_industry_category'].unique()

array(['Manufacturing', 'Property', 'Financial Services', 'Entertainment',
       'Retail', 'IT', 'Telecommunications', 'Health', nan, 'Argiculture'],
      dtype=object)

In [26]:
df_new_customer_list['wealth_segment'].unique()

array(['Mass Customer', 'Affluent Customer', 'High Net Worth'],
      dtype=object)

In [27]:
df_new_customer_list['owns_car'].unique()

array(['Yes', 'No'], dtype=object)

In [28]:
df_new_customer_list['state'].unique()

array(['QLD', 'NSW', 'VIC'], dtype=object)

In [29]:
df_new_customer_list['country'].unique()

array(['Australia'], dtype=object)

We can see the data is consistent.

### Data Containing Allowable Values - Validity

Data is valid

### Records that are Duplicated - Uniqueness

In [30]:
df_new_customer_list[df_new_customer_list.duplicated()]

Empty DataFrame
Columns: [first_name, last_name, gender, past_3_years_bike_related_purchases, DOB, job_title, job_industry_category, wealth_segment, deceased_indicator, owns_car, tenure, address, postcode, state, country, property_valuation, Unnamed: 16, Unnamed: 17, Unnamed: 18, Unnamed: 19, Unnamed: 20, Rank, Value]
Index: []

[0 rows x 23 columns]

No duplication found.

## Assess CustomerDemographic

In [31]:
df_customer_demographic.head()

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator  \
0                Health      Mass Customer                  N   
1    Financial Services      Mass Customer                  N   
2              Property      Mass Customer                  N   
3                    IT      Mass Customer                  N   
4                   NaN  Affluent Customer                  N   

                                             default owns_car  tenure  
0                                                 "'      Yes    11.0  
1                       <script>alert('hi')</script>      Yes    16.0  
2                                2018-02-01 00:00:00      Yes    15.0  
3  () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                NIL      Yes     8.0

### Correct Values - Accuracy

In [32]:
df_customer_demographic.describe()

customer_id  past_3_years_bike_related_purchases       tenure
count  4000.000000                          4000.000000  3913.000000
mean   2000.500000                            48.890000    10.657041
std    1154.844867                            28.715005     5.660146
min       1.000000                             0.000000     1.000000
25%    1000.750000                            24.000000     6.000000
50%    2000.500000                            48.000000    11.000000
75%    3000.250000                            73.000000    15.000000
max    4000.000000                            99.000000    22.000000

Data is accurate.

### Data Fields with Values - Completeness

In [33]:
df_customer_demographic.isnull().sum()

customer_id                              0
first_name                               0
last_name                              125
gender                                   0
past_3_years_bike_related_purchases      0
DOB                                     87
job_title                              506
job_industry_category                  656
wealth_segment                           0
deceased_indicator                       0
default                                302
owns_car                                 0
tenure                                  87
dtype: int64

We have multiple rows with null values. This need to be addressed with the client.

### Values Free from Contradition - Consistency

In [34]:
df_customer_demographic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 0 to 3999
Data columns (total 13 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   customer_id                          4000 non-null   int64         
 1   first_name                           4000 non-null   object        
 2   last_name                            3875 non-null   object        
 3   gender                               4000 non-null   object        
 4   past_3_years_bike_related_purchases  4000 non-null   int64         
 5   DOB                                  3913 non-null   datetime64[ns]
 6   job_title                            3494 non-null   object        
 7   job_industry_category                3344 non-null   object        
 8   wealth_segment                       4000 non-null   object        
 9   deceased_indicator                   4000 non-null   object        
 10  default     

Columns with dtype object should be assessed.

In [35]:
df_customer_demographic['gender'].unique()

array(['F', 'Male', 'Female', 'U', 'Femal', 'M'], dtype=object)

Gender is not consistent. For example, we have "F", "Female", "Femal" represent female. However, this can be cleaned by data wrangling.

In [36]:
df_customer_demographic['job_industry_category'].unique()

array(['Health', 'Financial Services', 'Property', 'IT', nan, 'Retail',
       'Argiculture', 'Manufacturing', 'Telecommunications',
       'Entertainment'], dtype=object)

df_customer_demographic['wealth_segment'].unique()

In [37]:
df_customer_demographic['deceased_indicator'].unique()

array(['N', 'Y'], dtype=object)

In [38]:
df_customer_demographic['owns_car'].unique()

array(['Yes', 'No'], dtype=object)

Besides gender, all other columns have good consistency.

### Data Containing Allowable Values - Validity

In [39]:
df_customer_demographic.head()

customer_id      first_name  last_name  gender  \
0            1         Laraine  Medendorp       F   
1            2             Eli    Bockman    Male   
2            3           Arlin     Dearle    Male   
3            4          Talbot        NaN    Male   
4            5  Sheila-kathryn     Calton  Female   

   past_3_years_bike_related_purchases        DOB               job_title  \
0                                   93 1953-10-12     Executive Secretary   
1                                   81 1980-12-16  Administrative Officer   
2                                   61 1954-01-20      Recruiting Manager   
3                                   33 1961-10-03                     NaN   
4                                   56 1977-05-13           Senior Editor   

  job_industry_category     wealth_segment deceased_indicator  \
0                Health      Mass Customer                  N   
1    Financial Services      Mass Customer                  N   
2              Property      Mass Customer                  N   
3                    IT      Mass Customer                  N   
4                   NaN  Affluent Customer                  N   

                                             default owns_car  tenure  
0                                                 "'      Yes    11.0  
1                       <script>alert('hi')</script>      Yes    16.0  
2                                2018-02-01 00:00:00      Yes    15.0  
3  () { _; } >_[$($())] { touch /tmp/blns.shellsh...       No     7.0  
4                                                NIL      Yes     8.0

In [40]:
df_customer_demographic['DOB'].sort_values()

33     1843-12-21
719    1931-10-23
1091   1935-08-22
3409   1940-09-22
2412   1943-08-11
          ...    
3778          NaT
3882          NaT
3930          NaT
3934          NaT
3997          NaT
Name: DOB, Length: 4000, dtype: datetime64[ns]

In [41]:
df_customer_demographic['DOB'].sort_values(ascending=False)

65     2002-03-11
1887   2002-01-26
3434   2002-01-15
2857   2002-01-09
421    2002-01-06
          ...    
3778          NaT
3882          NaT
3930          NaT
3934          NaT
3997          NaT
Name: DOB, Length: 4000, dtype: datetime64[ns]

We have a customer born on 1843-12-21. This is very unlikely. This need to be addressed with the client.

### Records that are Duplicated - Uniqueness

In [42]:
df_customer_demographic[df_customer_demographic.duplicated()]

Empty DataFrame
Columns: [customer_id, first_name, last_name, gender, past_3_years_bike_related_purchases, DOB, job_title, job_industry_category, wealth_segment, deceased_indicator, default, owns_car, tenure]
Index: []

The data ensures good uniqueness

## Assess CustomerAddress

In [43]:
df_customer_address.head()

customer_id              address  postcode            state    country  \
0            1   060 Morning Avenue      2016  New South Wales  Australia   
1            2  6 Meadow Vale Court      2153  New South Wales  Australia   
2            4   0 Holy Cross Court      4211              QLD  Australia   
3            5  17979 Del Mar Point      2448  New South Wales  Australia   
4            6     9 Oakridge Court      3216              VIC  Australia   

   property_valuation  
0                  10  
1                  10  
2                   9  
3                   4  
4                   9

### Correct Values - Accuracy

In [44]:
df_customer_address.describe()

customer_id     postcode  property_valuation
count  3999.000000  3999.000000         3999.000000
mean   2003.987997  2985.755939            7.514379
std    1154.576912   844.878364            2.824663
min       1.000000  2000.000000            1.000000
25%    1004.500000  2200.000000            6.000000
50%    2004.000000  2768.000000            8.000000
75%    3003.500000  3750.000000           10.000000
max    4003.000000  4883.000000           12.000000

Data is accurate

### Data Fields with Values - Completeness

In [45]:
df_customer_address.isnull().sum()

customer_id           0
address               0
postcode              0
state                 0
country               0
property_valuation    0
dtype: int64

Data is complete

### Values Free from Contradition - Consistency

In [46]:
df_customer_address.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 0 to 3998
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         3999 non-null   int64 
 1   address             3999 non-null   object
 2   postcode            3999 non-null   int64 
 3   state               3999 non-null   object
 4   country             3999 non-null   object
 5   property_valuation  3999 non-null   int64 
dtypes: int64(3), object(3)
memory usage: 187.6+ KB


In [47]:
df_customer_address['state'].unique()

array(['New South Wales', 'QLD', 'VIC', 'NSW', 'Victoria'], dtype=object)

In [48]:
df_customer_address['country'].unique()

array(['Australia'], dtype=object)

State is inconsistent. However, this can be cleaned by data wrangling.

### Data Containing Allowable Values - Validity

Data is valid

### Records that are Duplicated - Uniqueness

In [49]:
df_customer_address[df_customer_address.duplicated()]

Empty DataFrame
Columns: [customer_id, address, postcode, state, country, property_valuation]
Index: []

Data is unique

## In General

### Values Up to Date - Currency and Data Items with Value Meta-data - Relevancy

Lets see if customer data covers the customer ID in the transaction data.

In [50]:
df_transactions.sample()

transaction_id  product_id  customer_id transaction_date  online_order  \
1970            1971          45         1914       2017-12-18           1.0   

     order_status  brand product_line product_class product_size  list_price  \
1970     Approved  Solex     Standard        medium       medium      441.49   

      standard_cost  product_first_sold_date  
1970          84.99                  34071.0

In [51]:
df_new_customer_list.sample()

first_name last_name gender  past_3_years_bike_related_purchases  \
221       Kort    Disley   Male                                   66   

           DOB         job_title job_industry_category wealth_segment  \
221 1979-02-05  Technical Writer                Health  Mass Customer   

    deceased_indicator owns_car  ...  state    country  property_valuation  \
221                  N      Yes  ...    NSW  Australia                   9   

    Unnamed: 16 Unnamed: 17  Unnamed: 18  Unnamed: 19  Unnamed: 20  Rank  \
221        0.76        0.95       1.1875     1.009375          219   219   

     Value  
221  1.125  

[1 rows x 23 columns]

In [52]:
df_customer_demographic.sample()

customer_id first_name  last_name gender  \
2960         2961       Merv  Barradell   Male   

      past_3_years_bike_related_purchases        DOB          job_title  \
2960                                   56 1975-03-22  Assistant Manager   

     job_industry_category wealth_segment deceased_indicator  \
2960                Health  Mass Customer                  N   

               default owns_car  tenure  
2960  ã»(ï¿£âï¿£)ã»:*:      Yes    12.0

In [53]:
df_customer_address.sample()

customer_id                 address  postcode state    country  \
1703         1708  39846 Springs Crossing      2154   NSW  Australia   

      property_valuation  
1703                  10

In [54]:
df_trans_slim = df_transactions[['transaction_id', 'customer_id']]
df_trans_slim.head()

transaction_id  customer_id
0               1         2950
1               2         3120
2               3          402
3               4         3135
4               5          787

In [55]:
df_customer_demo_slim = df_customer_demographic[['customer_id']]
df_customer_demo_slim['has_demo_data'] = True
df_customer_demo_slim.head()

<ipython-input-55-7ef496c2a719>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_customer_demo_slim['has_demo_data'] = True


customer_id  has_demo_data
0            1           True
1            2           True
2            3           True
3            4           True
4            5           True

In [56]:
df_customer_add_slim = df_customer_address[['customer_id']]
df_customer_add_slim['has_add_data'] = True
df_customer_add_slim.head()

<ipython-input-56-0c15d089dc8a>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_customer_add_slim['has_add_data'] = True


customer_id  has_add_data
0            1          True
1            2          True
2            4          True
3            5          True
4            6          True

In [57]:
df_trans_demo_add_slim = df_trans_slim.merge(df_customer_demo_slim, how='left', on='customer_id')
df_trans_demo_add_slim = df_trans_demo_add_slim.merge(df_customer_add_slim, how='left', on='customer_id')
df_trans_demo_add_slim.head()

transaction_id  customer_id has_demo_data has_add_data
0               1         2950          True         True
1               2         3120          True         True
2               3          402          True         True
3               4         3135          True         True
4               5          787          True         True

In [58]:
df_trans_demo_add_slim.isnull().sum()

transaction_id     0
customer_id        0
has_demo_data      3
has_add_data      32
dtype: int64

In [59]:
df_trans_demo_add_slim[df_trans_demo_add_slim['has_demo_data'].isnull()]

transaction_id  customer_id has_demo_data has_add_data
8707             8708         5034           NaN          NaN
16700           16701         5034           NaN          NaN
17468           17469         5034           NaN          NaN

In [60]:
df_trans_demo_add_slim[df_trans_demo_add_slim['has_add_data'].isnull()]

transaction_id  customer_id has_demo_data has_add_data
117               118           22          True          NaN
3479             3480           22          True          NaN
3503             3504           22          True          NaN
4564             4565           10          True          NaN
4600             4601           22          True          NaN
5597             5598           23          True          NaN
5955             5956           10          True          NaN
7428             7429           23          True          NaN
7694             7695           23          True          NaN
7950             7951           23          True          NaN
8206             8207           22          True          NaN
8707             8708         5034           NaN          NaN
10249           10250           10          True          NaN
10301           10302            3          True          NaN
11004           11005            3          True          NaN
11783           11784            3          True          NaN
12953           12954            3          True          NaN
13284           13285            3          True          NaN
14898           14899           10          True          NaN
14955           14956           22          True          NaN
15187           15188            3          True          NaN
15586           15587           23          True          NaN
16094           16095           10          True          NaN
16625           16626           22          True          NaN
16700           16701         5034           NaN          NaN
16724           16725            3          True          NaN
17001           17002           22          True          NaN
17468           17469         5034           NaN          NaN
17697           17698           23          True          NaN
18218           18219           23          True          NaN
18945           18946           10          True          NaN
19497           19498            3          True          NaN

Customer demographic and address data is not up to date with transactions data, despite the fact that demographic data is more current that address data. This need to be addressed with the client.

Dataframes are relevant.

## Conclusion

'Transactions', 'NewCustomerList', and 'CustomerDemographic' have completeness issues. There are data enties with null values.

In 'CustomerDemographic' we have a invalid customer DOB (1843-12-21).

'CustomerDemographic' and 'CustomerAddress' are not up to date with transactions data.